desktop/cmp/dashGrid/DashGridLayoutContainerModel.js

```diff
@@ -0,0 +1,295 @@
+                initialValue: view.title,
+                rules: [required]
+            }
+        });
+        if (newName) view.setTitle(newName);
+    }
+
+    getViewSpec(id) {
+        return this.viewSpecs.find(it => it.id === id);
+    }
+
+    // Get all ViewModels with a given DashViewSpec.id
+    getItemsBySpecId(id) {
+        return this.viewModels.filter(it => it.viewSpec.id === id);
+    }
+
+    setInitialViews() {
+        this.removeAllViews();
+        this.initialViews.forEach(view => {
+            const {id, x, y, width, height} = view;
+            this.addView(id, {x, y, width, height});
+        });
+    }
+
+    /**
+     * Restore the initial state as specified by the application at construction time. This is the
+     * state without any persisted state or user changes applied.
+     *
+     * This method will clear the persistent state saved for this component, if any.
+     */
+    restoreDefaults() {
+        const {columns, rowHeight} = this.restoreState;
+        this.setColumns(columns);
+        this.setRowHeight(rowHeight);
+        this.setInitialViews();
+        this.provider?.clear();
+    }
+
+    //------------------------
+    // Implementation
+    //------------------------
+    /**
+     * Used to set the {x, y} position for the next added item based on the mouse location when
+     * context menu is triggered
+     * @param {number} x - clientX position
+     * @param {number} y - clientY position
+     */
+    setNextPosition(x, y) {
+        const calcXY = (positionParams, top, left, w=0, h=0) => {
+            const calcGridColWidth = (positionParams) => {
+                const { margin, containerPadding, containerWidth, cols } = positionParams;
+                return (
+                    (containerWidth - margin[0] * (cols - 1) - containerPadding[0] * 2) / cols
+                );
+            };
+
+            const clamp = (num, lowerBound, upperBound) =>
+                Math.max(Math.min(num, upperBound), lowerBound);
+
+            const {margin, cols, rowHeight, maxRows} = positionParams;
+            const colWidth = calcGridColWidth(positionParams);
+            let x = Math.round((left - margin[0]) / (colWidth + margin[0]));
+            let y = Math.round((top - margin[1]) / (rowHeight + margin[1]));
+
+            x = clamp(x, 0, cols - w);
+            y = clamp(y, 0, maxRows - h);
+            return { x, y };
+        };
+
+        const {margin, columns: cols, rowHeight, maxRows, ref, containerPadding} = this,
+            containerPosition = ref.current.getBoundingClientRect(),
+            {left: containerLeft, top: containerTop, width: containerWidth} = containerPosition,
+            positionParams = {margin, cols, rowHeight, maxRows,
+                containerPadding: containerPadding ?? margin, containerWidth},
+            left = x - containerLeft,
+            top = y - containerTop;
+
+        this.nextPosition = calcXY(positionParams, top, left);
```

Comment:
See other comment about potentially passing the click position through the context menu so we don't need to keep this temporary value in the model here, and could then calculate this next position at the point of adding the view from the ctx menu.